### DISCUSSION TITLE
get independent inputs (not switching the output) on ESP32 with MCP23017

### DISCUSSION THREAD

**Discussion by Eisbahn** [arendst/Tasmota#21689] [Q&A]
Dear all,

I'm using an ESP32 with several MCP23017 and was somehow lucky, it worked quite fine. The function and states are controlled via a server, connection is done via MQTT.
My config file looks like following, giving me (on three MCP23017 in a chain) 24 outputs and 21 inputs (hint: the MCP23017 allows PINs GPA7 and GPB7 only as output, hence in fact three less inputs/setting them to low) in total:
```
{"NAME":"MCP23017","BASE":0,"GPIO":[256,257,258,259,260,261,262,263,264,265,266,267,268,269,270,271,272,273,274,275,276,277,278,279,32,33,34,35,36,37,38,3872,40,41,42,43,44,45,46,3872,48,49,50,51,52,53,54,3872]}
```
This works fine.
BUT: every input is coupled directly with an output. Meaning if I press a button on one of the inputs, the corresponding output is changing its state, e.g. I press the first button once, the first output toggles from off to on, I press it a second time, it changes back to off. Nice feature, but I do want to just get the inputs forwarded via MQTT and decide on a server what should happen, e.g. switch two lights on if an input is detected, or if input no.3 is pressed, switch output six on or similar.
Is there any way to influence the internal couppling of the IOs and delete it? Otherwise I would need to overrule them on server side and immediately send via server the desired state, e.g. input 4 state changed received results in XOR to output 4 (to get the old state back) and switch output 2 on, output 10 off, output 2 for 5min to on (or whatever is needed).
Any hint available?

**Comment by Noschvie** [comment]
> hint: the MCP23017 allows PINs GPA7 and GPB7 only as output

is this a limitation of the MCP or of Tasmota?
Thanks

**Reply by Eisbahn** [reply]
> is this a limitation of the MCP or of Tasmota?

The chip itself, e.g. cmp. to
https://www.microchip.com/wwwproducts/en/MCP23017
https://ww1.microchip.com/downloads/aemDocuments/documents/APID/ProductDocuments/DataSheets/MCP23017-Data-Sheet-DS20001952.pdf

**Comment by lalo-uy** [comment]
You can decuple buttons and relays with switchmode cmd

https://tasmota.github.io/docs/Buttons-and-Switches/

**Reply by Eisbahn** [reply]
> You can decuple buttons and relays with switchmode cmd
> 
> https://tasmota.github.io/docs/Buttons-and-Switches/

I will try it tomorrow, however: I do have some doubts:
`SwitchMode, as the name implies, applies ONLY to GPIO
[...]
SwitchMode has NO impact on the behavior of GPIO configured as Button<x>`
I'm no on the GPIOs of the ESP. However, will try to set it to `switch mode 15` and `ButtonTopic 1`

**Reply by sfromis** [reply]
Not tried this, but with mode 2, you can have MCP23017 pins be treated as if they were "direct" gpio pins, Switch/Button. I'd "not be surprised" if this could work for SwitchMode too, or the direct `SetOption114` to decouple all Switch inputs. Or `SetOption73` to decouple all Button inputs. Better avoid `ButtonTopic` unless you have a need for this very specialized usage.

**Comment by Eisbahn** [comment]
Mode 2 is the only usable mode and used by me (see config in first post), as I'm using three of the expanders in a chain. Mode 1 supports only 1 expander -> so I'm on Mode 2 anyways.
SetOption73 decoupled it at least.
However: I don't get reliable MQTT info for each button press at the moment. 

I used ButtonTopic and set it to `ButtonTopic mybutton` to find it much more easy in MQTT sniffer. But: I still don't get messages each time (button is fine, voltage changed on pin on press)

**Reply by sfromis** [reply]
I'd say that using `ButtonTopic` gives you less info about what is happening, compared to using `SetOption73` only, as the latter also informs about what action Tasmota detected on the `Button` input. This may be the reason why you are not getting "reliable" info, as `ButtonTopic` hides info about features like multi or long press. It will only give the final result, instead of each push.

`ButtonTopic` is really only designed for a narrow use case, when you want to send Power commands to another Tasmota device, instead of having it act on the same device as having the buttons. But is only the commands, instead of full info. That's why I recommended to **not** use this feature.

Basically, you shot yourself in the foot, instead of getting easy info in your sniffer :grinning:
